File: go-web-basics/code-samples/servemux/servemux.go
package main

import (
	"fmt"
	"net/http"
)

func main() {
	mux := http.NewServeMux()

	//Matches ONLY /sloths, not /sloths/ or /sloths/are-awesome
	mux.HandleFunc("/sloths", func(w http.ResponseWriter, r *http.Request) {
		fmt.Fprintf(w, "Sloths rule!")
	})

	//Matches /kangaroos, /kangaroos/, and /kangaroos/tree-kangaroos
	mux.HandleFunc("/kangaroos/", func(w http.ResponseWriter, r *http.Request) {
		fmt.Fprintf(w, "Kangaroos for the win!")
	})

	//Since "/" ends with a slash, it matches all URL paths, so "/"
	//is the catch-all route.
	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		fmt.Fprintf(w, "Lemurs get the catch-all route! Lemurs are where it's at!")
	})

	// tea/hibiscus matches this route, but the route below is longer
	mux.HandleFunc("/tea/", func(w http.ResponseWriter, r *http.Request) {
		fmt.Fprintf(w, "One tea coming right up!")
	})

	//This route has a longer path so a request to tea/hibiscus matches this
	//route, not the one above.
	mux.HandleFunc("/tea/hibiscus", func(w http.ResponseWriter, r *http.Request) {
		fmt.Fprintf(w, "One hibiscus tea coming right up!")
	})

	server := &http.Server{
		Addr:    ":1123",
		Handler: mux,
	}
	server.ListenAndServe()
}

File: go-web-basics/code-samples/servemux/servemux.js
var http    = require('http'),
    express = require('express')

var app = express()

//tea/hibiscus matches this route
app.get('/tea(/*)?', function(req, res){
  res.send('One tea coming right up!')
})

//Since tea/hibiscus already matched the above route, this route
//isn't matched, so this route should be defined first if we want this
//route to be matched first.
app.get('/tea/hibiscus', function(req, res){
  res.send('One hibiscus tea coming right up!')
})

//Matches ONLY /sloths and /sloths/, not /sloths/are-awesome
app.get('/sloths', function(req, res){
  res.send('Sloths rule!')
})

//Matches /kangaroos, /kangaroos/, and /kangaroos/tree-kangaroos
app.get('/kangaroos(/*)?', function(req, res){
  res.send('Kangaroos for the win!')
})

//Since "*" matches all URL paths, so "*" is the catch-all route.
app.get('*', function(req, res){
  res.send("Lemurs get the catch-all route! Lemurs are where it's at!")
})

var server = http.createServer(app)
server.listen(1123)
